# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
# Use subject to license terms.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, and/or sell copies of the Software, and to permit persons
# to whom the Software is furnished to do so, provided that the above
# copyright notice(s) and this permission notice appear in all copies of
# the Software and that both the above copyright notice(s) and this
# permission notice appear in supporting documentation.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
# OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL
# INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING
# FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
# WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
# 
# Except as contained in this notice, the name of a copyright holder
# shall not be used in advertising or otherwise to promote the sale, use
# or other dealings in this Software without prior written authorization
# of the copyright holder.

6205507 - Xorg errors on ATI Radeon: xf86MapVidMem: failed to open /dev/xsvc (Not owner)


--- hw/xfree86/os-support/solaris/sun_vid.c	2006-10-11 13:46:15.000000000 -0700
+++ hw/xfree86/os-support/sunos/sun_vid.c	2005-11-03 14:13:07.732827000 -0800
@@ -39,6 +40,11 @@
 #define MAP_FAILED ((void *)-1)
 #endif
 
+#ifdef SUNSOFT
+static int apertureDevFD_ro = -1;
+static int apertureDevFD_rw = -1;
+#endif
+
 /***************************************************************************/
 /* Video Memory Mapping section 					   */
 /***************************************************************************/
@@ -71,7 +77,24 @@
 		return FALSE;
 	    }
 	}
+#ifdef SUNSOFT
+	apertureDevFD_rw = mmapFd;
+	if ((apertureDevFD_ro = open(apertureDevName, O_RDONLY)) < 0)
+	{
+	  xf86MsgVerb(X_WARNING, 0,
+		      "xf86LinearVidMem: failed to open %s (%s)\n",
+		      apertureDevName, strerror(errno));
+	  xf86MsgVerb(X_WARNING, 0,
+		      "xf86LinearVidMem: either /dev/fbs/aperture or /dev/xsvc"
+		      " device driver required\n");
+	  xf86MsgVerb(X_WARNING, 0,
+		      "xf86LinearVidMem: linear memory access disabled\n");
+	  apertureDevName = NULL;
+	  return FALSE;
+	}
+#else
 	close(mmapFd);
+#endif
 	return TRUE;
 }
 
@@ -119,8 +142,11 @@
 
 		strcpy(vtname, apertureDevName);
 	}
-
+#ifdef SUNSOFT
+	fd = dup((Flags & VIDMEM_READONLY) ? apertureDevFD_ro : apertureDevFD_rw);
+#else
 	fd = open(vtname, (Flags & VIDMEM_READONLY) ? O_RDONLY : O_RDWR);
+#endif
 	if (fd < 0)
 		FatalError("xf86MapVidMem: failed to open %s (%s)\n",
 			   vtname, strerror(errno));
